### PR TITLE
WEBUI-730/731: add designer catalog generator

### PR DESCRIPTION
**Pull request by richardsd**

### CODE REVIEW

**Review by poonamyadav252 — COMMENTED** [review]
It looks nice for now.

**Review by rakeshkumar1019 — COMMENTED** [review]
It looks good.

**Review by rakeshkumar1019 — DISMISSED** [review]

**Review by swarnadipa-dev — DISMISSED** [review]
Looks good

**Review by SnehalNaik-DEV — DISMISSED** [review]
Looks good

File: .github/workflows/catalog.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      target_platform:
+        description: 'The Nuxeo server target platform to build the catalog for.'
+        default: '11.3'
+        required: true
+      branch_name:
+        description: 'The name of the Nuxeo Web UI branch to generate the catalog for.'
+        default: 'maintenance-3.0.x'
+        required: true
+      classifier:
+        description: 'Classifier to generate a custom catalog version for testing purposes'
```
Comment (rakeshkumar1019) [inline .github/workflows/catalog.yaml]: here also missed `.`.
 description: 'Classifier to generate a custom catalog version for testing purposes`.`'

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      "description":"Nuxeo elements providing UI and common utility components for building custom applications."
+    },
+    {
+      "name":"nuxeo-dataviz-elements",
+      "title":"Dataviz Elements",
+      "symbol":"VZ",
+      "color":"#81c784",
+      "version":"1.0",
+      "tagline":"Nuxeo data visualization elements",
+      "description":"Nuxeo elements for building custom data visualization applications and dashboards"
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json:44]: here also 
"description":"Nuxeo elements for building custom data visualization applications and dashboards`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    },
+    {
+      "name":"nuxeo-web-ui",
+      "title":"Web UI Elements",
+      "symbol":"WUI",
+      "color":"#81c784",
+      "version":"master",
+      "tagline":"Web UI elements",
+      "description":"Nuxeo Web UI default elements",
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json:53]: "description":"Nuxeo Web UI default elements`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/11.1/catalog-packages.json [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      "description":"Nuxeo elements providing UI and common utility components for building custom applications."
+    },
+    {
+      "name":"nuxeo-dataviz-elements",
+      "title":"Dataviz Elements",
+      "symbol":"VZ",
+      "color":"#81c784",
+      "version":"1.0",
+      "tagline":"Nuxeo data visualization elements",
+      "description":"Nuxeo elements for building custom data visualization applications and dashboards"
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/11.1/catalog-packages.json:28]: "description":"Nuxeo elements for building custom data visualization applications and dashboards`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/11.1/catalog-packages.json [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    },
+    {
+      "name":"nuxeo-web-ui",
+      "title":"Web UI Elements",
+      "symbol":"WUI",
+      "color":"#81c784",
+      "version":"master",
+      "tagline":"Web UI elements",
+      "description":"Nuxeo Web UI default elements",
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/11.1/catalog-packages.json:37]: "description":"Nuxeo Web UI default elements`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/11.3/catalog-packages.json [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      "description":"Nuxeo elements providing UI and common utility components for building custom applications."
+    },
+    {
+      "name":"nuxeo-dataviz-elements",
+      "title":"Dataviz Elements",
+      "symbol":"VZ",
+      "color":"#81c784",
+      "version":"1.0",
+      "tagline":"Nuxeo data visualization elements",
+      "description":"Nuxeo elements for building custom data visualization applications and dashboards"
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/11.3/catalog-packages.json:28]: "description":"Nuxeo elements for building custom data visualization applications and dashboards`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/11.3/catalog-packages.json [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    },
+    {
+      "name":"nuxeo-web-ui",
+      "title":"Web UI Elements",
+      "symbol":"WUI",
+      "color":"#81c784",
+      "version":"master",
+      "tagline":"Web UI elements",
+      "description":"Nuxeo Web UI default elements",
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/11.3/catalog-packages.json:37]: "description":"Nuxeo Web UI default elements`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/8.10/catalog-packages.json [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      "description":"Nuxeo elements providing UI and common utility components for building custom applications."
+    },
+    {
+      "name":"nuxeo-dataviz-elements",
+      "title":"Dataviz Elements",
+      "symbol":"VZ",
+      "color":"#81c784",
+      "version":"1.0",
+      "tagline":"Nuxeo data visualization elements",
+      "description":"Nuxeo elements for building custom data visualization applications and dashboards"
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/8.10/catalog-packages.json:46]: "description":"Nuxeo elements for building custom data visualization applications and dashboards`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/8.10/catalog-packages.json [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    {
+      "name":"nuxeo-web-ui",
+      "title":"Web UI Elements",
+      "symbol":"WUI",
+      "color":"#81c784",
+      "version":"1.0",
+      "tagline":"Web UI elements",
+      "description":"Nuxeo Web UI default elements",
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/8.10/catalog-packages.json:55]: "description":"Nuxeo Web UI default elements`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/9.10/catalog-packages.json [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      "description":"Nuxeo elements providing UI and common utility components for building custom applications."
+    },
+    {
+      "name":"nuxeo-dataviz-elements",
+      "title":"Dataviz Elements",
+      "symbol":"VZ",
+      "color":"#81c784",
+      "version":"1.0",
+      "tagline":"Nuxeo data visualization elements",
+      "description":"Nuxeo elements for building custom data visualization applications and dashboards"
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/9.10/catalog-packages.json:44]: "description":"Nuxeo elements for building custom data visualization applications and dashboards`.`"

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/9.10/catalog-packages.json [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    },
+    {
+      "name":"nuxeo-web-ui",
+      "title":"Web UI Elements",
+      "symbol":"WUI",
+      "color":"#81c784",
+      "version":"master",
+      "tagline":"Web UI elements",
+      "description":"Nuxeo Web UI default elements",
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/data/applications/nuxeo/9.10/catalog-packages.json:53]: "description":"Nuxeo Web UI default elements`.`"

File: packages/nuxeo-designer-catalog/src/hints/generators/polymer-hint-generator.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  }
+  elements.forEach((element) => {
+    const hint = {};
+    hint.doc = typeof element.desc !== 'undefined' ? element.desc : null;
+    element.properties.forEach((property) => {
+      const { name } = property;
+      const doc = typeof property.desc !== 'undefined' ? property.desc : null;
+      if (!property.private) {
+        if (property.function) {
+          if (property.params && property.params.length > 0) {
```
Comment (rakeshkumar1019) [inline packages/nuxeo-designer-catalog/src/hints/generators/polymer-hint-generator.js]: Suggestion : we can check both the conditions  using  `Array.isArray()` function.
`if (property.params && property.params.length > 0)` to `if (Array.isArray(property.params))`
for Reference : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray

**Review by rakeshkumar1019 — CHANGES_REQUESTED** [review]

File: packages/nuxeo-designer-catalog/src/hints/generators/polymer-hint-generator.js (hunk 12 above)
Comment (richardsd) [reply packages/nuxeo-designer-catalog/src/hints/generators/polymer-hint-generator.js]: Yes we can, thanks @rakeshkumar1019

File: packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json (hunk 2 above)
Comment (richardsd) [reply packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json:44]: To keep it consistent with the current behavior we already have in Designer (where these files are loaded), I don't think we need to add the `.` in the `*.json` files.

Comment (rakeshkumar1019) [reply packages/nuxeo-designer-catalog/data/applications/nuxeo/10.10/catalog-packages.json:44]: Okay, thanks for letting me know.

**Review by rakeshkumar1019 — DISMISSED** [review]
looks good to me.

File: .github/workflows/catalog.yaml [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      target_platform:
+        description: 'The Nuxeo server target platform to build the catalog for.'
+        default: '11.3'
+        required: true
+      branch_name:
+        description: 'The name of the Nuxeo Web UI branch to generate the catalog for.'
+        default: 'maintenance-3.0.x'
+        required: true
+      classifier:
+        description: 'Classifier to generate a custom catalog version for testing purposes.'
```
Comment (Gabez0r) [inline .github/workflows/catalog.yaml]: Not sure what this is. Leaving a comment here could help.

**Review by Gabez0r — COMMENTED** [review]
LGTM

File: .github/workflows/catalog.yaml (hunk 13 above)
Comment (richardsd) [reply .github/workflows/catalog.yaml]: It is just a classifier used for naming the catalog package that is generated.
I can leave a comment here (or even remove it 🤔)

Comment (Gabez0r) [reply .github/workflows/catalog.yaml]: In what situation do we need to change this?

Comment (Gabez0r) [reply .github/workflows/catalog.yaml]: I'm ok with removing it if we don't plan on using it, or leaving a comment like "change this if you need to ..." for clarity

Comment (richardsd) [reply .github/workflows/catalog.yaml]: This is used for testing. Imagine that you want to build a custom version of the catalog named `3.0.x-my-cool-catalog`, where the `my-cool-catalog` is the classifier

Comment (Gabez0r) [reply .github/workflows/catalog.yaml]: It seems useful. Leaving that on a comment (or description even?) would help.

File: packages/nuxeo-designer-catalog/scripts/maven-deploy.sh [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+ROOT=$1
+VERSION=$2
+GROUPID=org.nuxeo.web.ui.studio
+ARTIFACTID=view-designer-catalog
+
+MVN_REPO_ID=public-releases
+MVN_REPO_URL=http://mavenin.nuxeo.com/nexus/content/repositories/public-releases
+if [[ $VERSION == *-SNAPSHOT ]]; then
+    MVN_REPO_ID=public-snapshots
+    MVN_REPO_URL=http://mavenin.nuxeo.com/nexus/content/repositories/public-snapshots
```
Comment (richardsd) [inline packages/nuxeo-designer-catalog/scripts/maven-deploy.sh]: It should be updated to `https://packages.nuxeo.com` and `maven-public`

**Review by Gabez0r — APPROVED** [review]

**Comment by richardsd** [comment]
Example PR generated by the workflow: https://github.com/nuxeo/nos/pull/1263

**Review by mnixo — APPROVED** [review]

**Pull request merged**